### PR TITLE
Filtered out non data-nodes in relevant cat api

### PR DESCRIPTION
**Pull request by nirmalc**
Filtered out non data nodes for fielddata and allocation cat api calls

https://github.com/elasticsearch/elasticsearch/issues/9214

### CODE REVIEW

**Comment by dadoonet** [comment]
Wondering if this should be an option.
Also I think this change should be documented.
Unsure but wondering if tests are passing (REST tests).

**Comment by clintongormley** [comment]
Hi @nirmalc 

Thanks for the PR.  Please could you add the relevant documentation and a test for this change, and we'll review.

thanks

**Comment by nirmalc** [comment]
will do that - need to figure out where are the REST tests

**Comment by nirmalc** [comment]
@dadoonet rest tests are passing as they check output/input formats in which there is no change.

**Comment by clintongormley** [comment]
@bleskes ok to merge?

**Comment by bleskes** [comment]
+1 to the change. Implementation wise I would just change the default of the nodes param to be:

```
   final String[] nodes = Strings.splitStringByCommaToArray(request.param("nodes", "data:true)); 
```

This will also give us the flag @dadoonet requested (`nodes=*`)

We still need a test before we can merge. The rest tests are run against the global cluster which means they will need to be run a couple of times to run into configuration that has client nodes. Let me know if you need help here. (see https://github.com/elasticsearch/elasticsearch/blob/master/TESTING.asciidoc#testing-the-rest-layer )

**Comment by s1monw** [comment]
@nirmalc ping :) are you coming back to this? I think we only need to make @bleskes happy here and I think the change it a pretty good addition!!

**Comment by s1monw** [comment]
@clintongormley I assigned this to you

**Comment by nirmalc** [comment]
@s1monw  , @clintongormley  - sorry ! forgot about this one , will take care of this today

**Comment by s1monw** [comment]
@nirmalc nothing to be sorry about! thanks for coming back to us!! :+1:

**Comment by clintongormley** [comment]
@bleskes changes pushed. OK to merge?

**Comment by nirmalc** [comment]
@bleskes any more changes needed from my side ?

**Comment by bleskes** [comment]
@nirmalc sorry for dropping this.. LGTM.  I tend to pull it in to 2.0 only as it is a breaking change. I'll let @clintongormley @s1monw and others a day or two to object and I'll pull it in.

File: docs/reference/cat/allocation.asciidoc [hunk 1]
```diff
@@ -2,7 +2,7 @@
 == cat allocation
 
 `allocation` provides a snapshot of how shards have located around the
-cluster and the state of disk usage.
+cluster and the state of disk usage. This call lists only data nodes in cluster by default.
```
Comment (clintongormley) [inline docs/reference/cat/allocation.asciidoc]: This implies that you can change the default.  How about:

    `allocation` provides a snapshot of how many shards are allocated to each data node
    and how much disk space they are using.

File: docs/reference/cat/fielddata.asciidoc [hunk 2]
```diff
@@ -2,7 +2,7 @@
 == cat fielddata
 
 `fielddata` shows information about currently loaded fielddata on a per-node
-basis.
+basis on all data nodes in cluster.
 
 [source,shell]
```
Comment (clintongormley) [inline docs/reference/cat/fielddata.asciidoc]: `fielddata` shows how much heap memory is currently being used by fielddata 
    on every data node in the cluster.

File: rest-api-spec/test/cat.allocation/10_basic.yaml [hunk 3]
```diff
@@ -61,9 +61,9 @@
               ( \s*                          #allow leading spaces to account for right-justified text
                 \d+                    \s+
                 \d+(\.\d+)?[kmgt]?b    \s+
-                (\d+(\.\d+)?[kmgt]b   \s+)?  #no value from client nodes
-                (\d+(\.\d+)?[kmgt]b   \s+)?  #no value from client nodes
-                (\d+                  \s+)?  #no value from client nodes
+                (\d+(\.\d+)?[kmgt]b   \s+)  #always should return value since we filter out non data nodes by default
```
Comment (clintongormley) [inline rest-api-spec/test/cat.allocation/10_basic.yaml]: drop "by default"

File: rest-api-spec/test/cat.allocation/10_basic.yaml [hunk 4]
```diff
@@ -61,9 +61,9 @@
...
+                (\d+(\.\d+)?[kmgt]b   \s+)  #always should return value since we filter out non data nodes by default
+                (\d+                  \s+)  #always should return value since we filter out non data nodes by default
                 [-\w.]+                \s+
```
Comment (clintongormley) [inline rest-api-spec/test/cat.allocation/10_basic.yaml]: drop "by default"

File: rest-api-spec/test/cat.allocation/10_basic.yaml [hunk 5]
```diff
@@ -111,6 +111,37 @@
             $/
 
 ---
+
+"All Nodes":
+
```
Comment (clintongormley) [inline rest-api-spec/test/cat.allocation/10_basic.yaml]: Have I misunderstood?  Can you remove the data:true filter?

**Comment by clintongormley** [comment]
Let me know if I've misunderstood the data:true option - if that can be disabled then I'll revise my comments.

**Comment by s1monw** [comment]
LGTM too

File: rest-api-spec/test/cat.allocation/10_basic.yaml (hunk 5 above)
Comment (nirmalc) [inline rest-api-spec/test/cat.allocation/10_basic.yaml]: data:true is by default , but overridable by any node selector such as nodes=*

**Comment by bleskes** [comment]
@clintongormley the _cat/allocation and field data api allows you to change the set of nodes you get info for via the `nodes=` url param. This PR only changes the default value from `*` to `data:true` . You can very easily switch back. Does that help?

**Comment by clintongormley** [comment]
thanks @nirmalc & @bleskes - I've revised my comments. just two minor doc changes

thanks

**Comment by clintongormley** [comment]
@bleskes please add version labels to this

**Comment by bleskes** [comment]
woops. Added.

**Comment by karmi** [comment]
Hi @nirmalc, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/9287.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by karmi** [comment]
@nirmalc, apologies for the notification above, this is a fault from some manual testing... please ignore it.

**Pull request closed without merging**